File: Session7/Java/GroupAnagrams.java
https://leetcode.com/problems/group-anagrams/

class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        
        Map<String, List<String>> anagramGroups = new HashMap<>();

        for (String str : strs) {
            int[] charCount = new int[26];
            for (char c : str.toCharArray()) {
                charCount[c - 'a']++;
            }
            StringBuilder key = new StringBuilder();
            for (int i = 0; i < 26; i++) {
                key.append("#").append(charCount[i]);
            }

            anagramGroups.computeIfAbsent(key.toString(), k -> new ArrayList<>()).add(str);
        }

        return new ArrayList<>(anagramGroups.values());
        
    }
}

File: Session7/C++/GroupAnagrams.cpp
https://leetcode.com/problems/group-anagrams/

class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        
        unordered_map<string, vector<string>> anagramGroups;
        
        for(auto str:strs) {
            int charCount[26] = {0};
            for(char c:str) {
                charCount[c - 'a']++;
            }
            string key;
            for(int i=0; i<26; i++) {
                key += "#" + to_string(charCount[i]);
            }

            anagramGroups[key].push_back(str);
        }
        
        vector<vector<string>> result;
        for(auto group:anagramGroups) {
            result.push_back(group.second);
        }
        
        return result;
        
    }
};